python/python/bystro/proteomics/handler.py

```diff
@@ -0,0 +1,362 @@
+    elif isinstance(sub, str):
+        return sub
+    else:
+        raise TypeError(sub)
+        return str(sub)
+
+
+def _to_string_or_empty_field_char(x: str | None, delims: dict[str, str]) -> str:
+    return str(x) if x is not None else delims["empty_field"]
+
+
+from typing import cast
+
+
+def _process_rows(resp: dict, field_names: list[str]) -> list[Row]:
+    parent_fields, child_fields = _get_header(field_names)
+    try:
+        discordant_idx = field_names.index("discordant")
+    except ValueError as val_err:
+        err_msg = f"response: {resp} with field_names: {field_names} lacked field: 'discordant'"
+        raise ValueError(err_msg) from val_err
+    rows: list[Row] = []
+    docs = resp["hits"]["hits"]
+    for doc in docs:
+        row = []
+        for parent_field, child_field in zip(parent_fields, child_fields):
+            row.append(_populate_data(child_field, doc["_source"].get(parent_field)))
+        row[discordant_idx][0][0] = cast(str, _fix_discordancy(row[discordant_idx][0][0]))  # type: ignore
+        rows.append(row)  # type: ignore
+    return rows
+
+
+def _fix_discordancy(x: bool | str) -> str:
+    return {False: "0", True: "1", "0": "0", "1": "1"}[x]
+
+
+# @ray.remote
+# def _process_query(
+#     query_args: dict,
+#     search_client_args: dict,
+#     field_names: list,
+#     chunk_output_name: str,
+#     delimiters,
+# ):
+#     """Process OpenSearch query and save results.
+
+#     - Run OpenSearch query against index specified in
+#     - search_client_args Write output to disk under chunk_output_name.
+#     - Warning: this method does not implement exception-based error
+#       handling but returns an error code of RAY_ERROR if any exception is encountered.
+#     - returns... what? TK
+#     """
+
+#     if resp["hits"]["total"]["value"] == 0:
+
+#     # Each sliced scroll chunk should get all records for that chunk
+
+#         with gzip.open(chunk_output_name, "wb") as fw:
+
+
+@ray.remote
+def _process_query_pure(
+    query_args: dict,
+    search_client_args: dict,
+    field_names: list,
+    delimiters: dict[str, str],
+) -> int | bytes:
+    """Process OpenSearch query and return results."""
+    client = OpenSearch(**search_client_args)
+    resp = client.search(**query_args)
+    return _process_response(resp, field_names, delimiters)
+
+
+def _process_response(resp, field_names, delimiters) -> int | bytes:
+    if resp["hits"]["total"]["value"] == 0:
+        return 0
+
+    # Each sliced scroll chunk should get all records for that chunk
+    if len(resp["hits"]["hits"]) != resp["hits"]["total"]["value"]:
+        err_msg = (
+            f"response hits: {len(resp['hits']['hits'])} didn't equal "
+            f"total value: {resp['hits']['total']['value']}. "
+            "This is a bug."
+        )
+        raise ValueError(err_msg)
+    rows = _process_rows(resp, field_names)
+    output_string = _make_output_string(rows, delimiters)
```

Comment:
I wouldn't do this. Just return the object. If you're going to be parsing it anyway, no need to resplit later. Ray will happily pickle values.